File: checks/ruby/lib/section508/checks/link_purpose.rb
# frozen_string_literal: true

module Section508
  module Checks
    # The link name must say where it goes.
    class LinkPurpose
      WEAK = /^(click here|here|read more|more|learn more|click|link|this link)$/i

      def self.link_name(link)
        name = Dom.norm(link)
        return name unless name.empty?

        Dom.attr(link, "aria-label").to_s.gsub(/\s+/, " ").strip
      end
      private_class_method :link_name

      def self.call(root)
        return [] unless root
        return [] unless root.respond_to?(:css)

        links = Dom.elements(root, "a[href]")
        return ["This section has no link."] if links.empty?

        links.each do |link|
          name = link_name(link)
          return ["This link has no text name."] if name.empty?
          return ["“#{name}” does not name the destination."] if WEAK.match?(name)
        end
        []
      end
    end
  end
end

File: checks/python/src/section508/checks/link_purpose.py
import re

from section508.dom import SPACE, select

WEAK = re.compile(
    r"^(click here|here|read more|more|learn more|click|link|this link)$",
    re.I,
)


def link_name(link):
    name = SPACE.sub(" ", link.get_text()).strip()
    if name:
        return name
    return SPACE.sub(" ", (link.get("aria-label") or "")).strip()


def check(root):
    if not root:
        return []

    links = select(root, "a[href]")
    if not links:
        return ["This section has no link."]

    for link in links:
        name = link_name(link)
        if not name:
            return ["This link has no text name."]
        if WEAK.match(name):
            return [f"“{name}” does not name the destination."]
    return []
